### PR TITLE
flattening/arrays.cpp: replace assert and throw with invariants

### PR DESCRIPTION
**Pull request by smowton**

### CODE REVIEW

**Comment by martin-cs** [comment]
On Wed, 2017-06-21 at 02:13 -0700, Chris Smowton wrote:
> Also replace assert(false) with throw while I'm there.

Could they be DATA_INVARIANTs instead because by the look of it they
are.

**Comment by smowton** [comment]
@martin-cs done

File: src/solvers/flattening/arrays.cpp [hunk 1]
```diff
@@ -49,11 +48,13 @@ literalt arrayst::record_array_equality(
...
-    std::cout << equality.pretty() << '\n';
-    throw "record_array_equality got equality without matching types";
+    prop.error() << equality.pretty() << messaget::eom;
+    DATA_INVARIANT(false, "record_array_equality got equality without matching types");
   }
 
-  assert(ns.follow(op0.type()).id()==ID_array);
+  DATA_INVARIANT(
+    ns.follow(op0.type()).id()==ID_array,
+    "record_array_equality parameter should be array-typed");
```
Comment (martin-cs) [inline src/solvers/flattening/arrays.cpp]: This is a lovely example of how to convert asserts and how to add meaningful information during the conversion.

**Comment by smowton** [comment]
https://github.com/diffblue/cbmc/pull/1047 needs to go first

File: src/solvers/flattening/arrays.cpp [hunk 2]
```diff
@@ -185,9 +189,10 @@ void arrayst::collect_arrays(const exprt &a)
   }
   else if(a.id()==ID_member)
   {
-    if(to_member_expr(a).struct_op().id()!=ID_symbol)
-      throw
-        "unexpected array expression: member with `"+a.op0().id_string()+"'";
+    DATA_INVARIANT(
+      to_member_expr(a).struct_op().id()==ID_symbol,
+      ("unexpected array expression: member with `"+
+       a.op0().id_string()+"'").c_str());
```
Comment (martin-cs) [inline src/solvers/flattening/arrays.cpp]: [Not an issue for this PR, my problem not yours] This looks a little clunky.  Do you think we could improve invariant.h to de-clunky this?

File: src/solvers/flattening/arrays.cpp [hunk 3]
```diff
@@ -200,20 +205,24 @@ void arrayst::collect_arrays(const exprt &a)
   else if(a.id()==ID_byte_update_little_endian ||
           a.id()==ID_byte_update_big_endian)
   {
-    assert(0);
+    DATA_INVARIANT(
+      false,
+      "byte_update should be removed before collect_arrays");
```
Comment (martin-cs) [inline src/solvers/flattening/arrays.cpp]: Perfect -- this is exactly why these should be converted manually.

**Comment by martin-cs** [comment]
On Wed, 2017-06-21 at 04:11 -0700, Chris Smowton wrote:
> https://github.com/diffblue/cbmc/pull/1047 needs to go first

Similar comments apply :-)

File: src/solvers/flattening/arrays.cpp (hunk 2 above)
Comment (smowton) [inline src/solvers/flattening/arrays.cpp]: It would also be good to make this lazy for cases like this which build an error string. How about something like:
```
#define INVARIANT(CONDITION, REASON) \
  do { \
    if(!(CONDITION)) \
    { \
      std::string reason = (REASON); \
      invariant_failed(__FILE__, __FUNCTION__, __LINE__, reason); \
    } \
  } while(0);
```

Comment (smowton) [inline src/solvers/flattening/arrays.cpp]: (If invariant_failed (currently check_invariant) throws then it should take a std::string not a const char* otherwise the c_str() will outlive the string it's taken from)

Comment (reuk) [inline src/solvers/flattening/arrays.cpp]: I think that looks good. It also gives the option to use an immediately-evaluated lambda as the `REASON` parameter, in the case that `ostringstream` is required to build the message string.

As you said, depending on the implementation of `invariant_failed` it might be better to do
```c++
std::string reason = (REASON); \
invariant_failed(__FILE__, __FUNCTION__, __LINE__, std::move(reason)); \
```

**Comment by smowton** [comment]
There's already an option for INVARIANT to throw rather than `abort()` as is I believe the default thing to do at present.

@martin-cs how about using `backtrace()` to take the trace at the throw site, then dumping it (as the standard logic_error.what() string) if it isn't caught?

**Comment by reuk** [comment]
My point is that currently the code *always* throws, but `INVARIANT` *may* throw depending on compile options. If existing code expects that inner functions may throw, `INVARIANT` will break those functions.

**Comment by reuk** [comment]
This is why I think having multiple behaviours for `INVARIANT` is a bad idea. People will start to depend on a particular implementation, and their code will be broken when different compile options are used.

**Comment by smowton** [comment]
Allowing the possibility to catch is desirable: agreed, though this requires anyone writing INVARIANT to be aware that their destructor might be about to get called, so they need to do bare minimum cleanup first. That seems like something they ought to explicitly authorise at the callsite. Whether people do this for each callsite will presumably depend on where the anomaly sits on the spectrum of "double-check that's supposed to cause a big explosion on the test bench and make it clear I've screwed up" to "corner case that I think is impossible but is conceivable under obscure input or incomplete understanding on my part". In particular I don't think it's likely the original author here intended the throws and the asserts to be divided up on that basis.

Also given the state of error handling in cbmc, I think it's very unlikely those throws had any intended catcher, except perhaps to print the error before themselves aborting, though I will check this tomorrow morning.

Basically I think I've talked myself into keeping the distinction between fatal (don't even call my destructor, abort) and nonfatal (I can't continue but destroying this object and carrying on is acceptable) invariants, determined at the callsite not with a compile option. In this particular case I would keep them fatal for now since I don't know off hand whether an instance of a prop_convt-derived converter can be sensibly interacted with after aborting halfway through, but I will check that nobody is actually catching these throws.

@martin-cs, thoughts?

**Comment by peterschrammel** [comment]
The distinction to be made is whether the code expects an error to be recoverable or non-recoverable. The asserts in the code have always been considered as non-recoverable errors. Hence, the action is to print the stack trace and abort (or alternatively throw a non_recoveable_error exception that is only to be caught on top-level by convention -- we should fix the behaviour to either the one or the other). 
A reason why someone might want to catch non-recoverable errors is when they occur in library code that cannot be easily changed. But this is not the case for us, so, catching non-recoverable errors would just be a bad workaround for a situation where a non-recoverable error should be explicitly made recoverable (which obviously requires additional work in the place where the error occurs).

**Comment by martin-cs** [comment]
On Wed, 2017-06-21 at 10:24 -0700, Chris Smowton wrote:
> @martin-cs, thoughts?

I think everyone has heard more than enough of my thoughts about
this :-\

(I am assuming that this is now sorted and no further actual comment is
needed)
